File: c++/c12_3_applications.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c12_3_applications.h"

namespace {
	bool finding_cycles2(const vector<vector<ll>>& adj) {
		for (const auto& c : connected(adj)) {
			ll sum = 0;
			for (ll i : c)
				sum += ssize(adj[i]);
			if (sum != 2 * ssize(c) - 2)
				return true;
		}
		return false;
	}

	optional<vector<bool>> bipartiteness_check(const vector<vector<ll>>& adj) {
		ll n = ssize(adj);
		vector<bool> visited(n);
		vector<bool> color(n);
		auto dfs = [&](this auto self, ll s, bool c) {
			if (visited[s])
				return color[s] == c;
			visited[s] = true;
			color[s] = c;
			for (ll u : adj[s]) {
				if (!self(u, !c)) {
					return false;
				}
			}
			return true;
			};
		for (ll i = 1; i < n; ++i) {
			if (!visited[i] && !dfs(i, false)) {
				return nullopt;
			}
		}
		return color;
	}

	multiset<multiset<ll>> convert(const vector<vector<ll>>& vec) {
		return vec | views::transform(ranges::to<multiset>()) | ranges::to<multiset>();
	}
}

vector<vector<ll>> connected(const vector<vector<ll>>& adj) {
	ll n = ssize(adj);
	vector<bool> visited(n);
	vector<vector<ll>> ret;
	auto dfs = [&](this auto self, ll s) {
		if (visited[s]) return;
		visited[s] = true;
		ret.back().push_back(s);
		for (ll u : adj[s])
			self(u);
		};
	for (ll i = 1; i < n; ++i)
		if (!visited[i]) {
			ret.resize(ret.size() + 1);
			dfs(i);
		}
	return ret;
}

bool finding_cycles(const vector<vector<ll>>& adj) {
	ll n = ssize(adj);
	vector<bool> visited(n);
	auto dfs = [&](this auto self, ll s, ll e) {
		if (visited[s])
			return true;
		visited[s] = true;
		for (ll u : adj[s]) {
			if (u != e && self(u, s)) {
				return true;
			}
		}
		return false;
		};
	for (ll i = 1; i < n; ++i) {
		if (!visited[i] && dfs(i, 0)) {
			return true;
		}
	}
	return false;
}

TEST(C123Applications, connected) {
	EXPECT_EQ(convert(connected({ {} })), multiset<multiset<ll>>{});
	EXPECT_EQ(convert(connected({ {},{} })), multiset<multiset<ll>>{{1}});
	EXPECT_EQ(convert(connected({ {},{3,4},{5},{1,4},{1,3},{2} })), (multiset<multiset<ll>>{ {1,3,4},{2,5} }));
}

TEST(C123Applications, findingCycles) {
	EXPECT_FALSE(finding_cycles({ {},{} }));
	EXPECT_TRUE(finding_cycles({ {},{3,4},{3,5},{1,2,4,5},{1,3},{2,3} }));
	EXPECT_TRUE(finding_cycles({ {},{},{3,4},{4,2},{2,3} }));
}

TEST(C123Applications, findingCycles2) {
	EXPECT_FALSE(finding_cycles2({ {},{} }));
	EXPECT_TRUE(finding_cycles2({ {},{3,4},{3,5},{1,2,4,5},{1,3},{2,3} }));
	EXPECT_TRUE(finding_cycles2({ {},{},{3,4},{4,2},{2,3} }));
}

TEST(C123Applications, bipartitenessCheck) {
	EXPECT_EQ(bipartiteness_check({ {} }), vector<bool>{false});
	EXPECT_EQ(bipartiteness_check({ {},{} }), (vector<bool>{false,false}));
	EXPECT_EQ(bipartiteness_check({ {},{2,4},{1,3,5},{2,5},{1,5},{2,3,4} }), nullopt);
	EXPECT_EQ(bipartiteness_check({ {},{},{3,4},{4,2},{2,3} }), nullopt);
}

File: python/c12_3_applications.py
import unittest
import sys

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def connected(adj):
    n=len(adj)
    visited=[False]*n
    def dfs(s):
        if visited[s]:
            return
        visited[s]=True
        yield s
        for u in adj[s]:
            yield from dfs(u)
    for i in range(1,n):
        if not visited[i]:
            yield [*dfs(i)]

def finding_cycles(adj):
    n=len(adj)
    visited=[False]*n
    def dfs(s,e):
        if visited[s]:
            return True
        visited[s]=True
        return any(dfs(u,s) for u in adj[s] if u!=e)
    for i in range(1,n):
        if not visited[i] and dfs(i,0):
            return True
    return False

def finding_cycles2(adj):
    return any(sum(len(adj[i]) for i in c)!=len(c)*2-2 for c in connected(adj))

def bipartiteness_check(adj):
    n=len(adj)
    visited=[False]*n
    color=[False]*n
    def dfs(s,c):
        if visited[s]:
            return color[s]==c
        visited[s]=True
        color[s]=c
        return all(dfs(u,not c) for u in adj[s])
    for i in range(1,n):
        if not visited[i] and not dfs(i,False):
            return None
    return color

class Test(unittest.TestCase):
    def test(self):
        self.assertCountEqual(map(sorted,connected([[]])),[])
        self.assertCountEqual(map(sorted,connected([[],[]])),[[1]])
        self.assertCountEqual(map(sorted,connected([[],[3,4],[5],[1,4],[1,3],[2]])),[[1,3,4],[2,5]])
        self.assertFalse(finding_cycles([[]]))
        self.assertTrue(finding_cycles([[],[3,4],[3,5],[1,2,4,5],[1,3],[2,3]]))
        self.assertTrue(finding_cycles([[],[],[3,4],[4,2],[2,3]]))
        self.assertFalse(finding_cycles2([[]]))
        self.assertTrue(finding_cycles2([[],[3,4],[3,5],[1,2,4,5],[1,3],[2,3]]))
        self.assertTrue(finding_cycles2([[],[],[3,4],[4,2],[2,3]]))
        self.assertEqual(bipartiteness_check([[]]),[False])
        self.assertEqual(bipartiteness_check([[],[]]),[False,False])
        self.assertIsNone(bipartiteness_check([[],[2,4],[1,3,5],[2,5],[1,5],[2,3,4]]))
        self.assertIsNone(bipartiteness_check([[],[],[3,4],[4,2],[2,3]]))

if __name__=='__main__':
    unittest.main()
